src/interfaces/csharp/swig_typemaps.i

```diff
@@ -285,51 +281,45 @@ TYPEMAP_STRINGFEATURES(float64_t, double, double)
 		return $null;
 	}
 
-	shogun::SGString<char>* strings=SG_MALLOC(shogun::SGString<char>, size_$1);
+	strings.reserve(size_$1);
 
 	for (i = 0; i < size_$1; i++) {
 		str = $input[i];
 		len = strlen(str);
-		max_len = shogun::CMath::max(len, max_len);
 
-		strings[i].slen = len;
-		strings[i].string = NULL;
+		strings.emplace_back(len);
 
-		if (len > 0) {
-			strings[i].string = SG_MALLOC(char, len);
-			sg_memcpy(strings[i].string, str, len);
-		}
+		sg_memcpy(strings.back().vector, str, len);
 	}
+}
 
-	SGStringList<char> sl;
-	sl.strings = strings;
-	sl.num_strings = size_$1;
-	sl.max_string_length = max_len;
-	$1 = sl;
+%typemap(freearg) std::vector<shogun::SGVector<char>> {
+	typemap_utils::free_if_pointer($1);
 }
 
-%typemap(out) shogun::SGStringList<char> {
-	shogun::SGString<char>* str = $1.strings;
-	int32_t i, j;
-	int32_t size = $1.num_strings;
-	int32_t max_size = 32;
+%typemap(out) std::vector<shogun::SGVector<char>> {
+	auto& strings = typemap_utils::cast_deref<std::vector<shogun::SGVector<char>>>($1);
+	int32_t size = strings.size();
+	int32_t max_size = 0;
+	for(auto& str : strings)
+		max_size = std::max(max_size, str.vlen);
 
 	char ** res = SG_MALLOC(char*, size + 1);
 	res[0] = SG_MALLOC(char, max_size);
 	sprintf(res[0], "%d", size);
-
-	for (i = 0; i < size; i++) {
-		res[i + 1] = SG_MALLOC(char, str[i].slen + 1);
-		sg_memcpy(res[i + 1], str[i].string, str[i].slen * sizeof(char));
+	
+	for (int i = 0; i < size; i++) {
+		res[i + 1] = SG_MALLOC(char, strings[i].vlen + 1);
+		sg_memcpy(res[i + 1], strings[i].vector, strings[i].vlen * sizeof(char));
 		
 		// null terminate string as C#'s Marshal.PtrToStringAnsi expects that
-		res[i+1][str[i].slen] = '\0';
+		res[i+1][strings[i].vlen] = '\0';
```

Comment:
this causes me some trouble.
Whereas previously, we had the semantic idea of string-lists, this conversion to strings made sense.
But not we have a vector of SGVectors, why would be convert this to continuous style strings -- in csharp, arrays of chars and strings are different types right?
I didn't think of this problem before.... what are your thoughts on this?

Reply:
Maybe the solution would be to use `std::vector<std::string>` for collections of character strings?

Reply:
I'm not sure I understand. We can't in a simple manner replace `std::vector<SGVector<char>>` with `std::vector<std::string>` as they have different semantics. Most of the functions (and classes) that accept `std::vector<SGVector<T>>` will be specialized for strings.

Reply:
Java also is the same. `int[][]` is used for example for string features of ints, while `String[]` for chars. Maybe be consistent with all and use `char[][]`? But then it's harder for the user.

Reply:
I also think that's why we should somehow separate between `SGVector` and `SGString`. We can create an `SGString` class that inherits from `SGVector` and has the functionality of returning/accepting std::strings if `T` is `char`, so that this convenience maps to c++ users as well.

Reply:
> I'm not sure I understand. We can't in a simple manner replace `std::vector<SGVector<char>>` with `std::vector<std::string>` as they have different semantics. Most of the functions (and classes) that accept `std::vector<SGVector<T>>` will be specialized for strings.

yes, that's exactly what I mean: here we convert SGVector<char> to String, which is kinda mixing sutff.

Reply:
> I also think that's why we should somehow separate between `SGVector` and `SGString`. We can create an `SGString` class that inherits from `SGVector` and has the functionality of returning/accepting std::strings if `T` is `char`, so that this convenience maps to c++ users as well.

mmh that might be an indea.

Or, I also also the other comment you made: users dont have to see the internals.
From an internals perspective, there is no point is representing anything as std::string, but rather SGVector<char>.
So we can

* map `SGVector<char>` to the native string type in the typemaps (currently done). Maybe not the nicest though convenient for users
* map to the native array types instead (array of char), and then offer some interface to extract string representations. This could be done at the `StringFeatres` level, i.e. `get_strings` (or different name) returns `std::vector<std::string>` but this method only works for `SGVector<char>`. And then `get_vectors` would return `std::vector<SGVector<...>>`. This way we could have clean semantics in internals (and typemaps) and wouldnt need to append `\0` at the end. Instead, all this would be happening inside our new getter. Get my point?

Reply:
Certainly we should have some conversion methods from SGVector<char> to std::string as well

Reply:
use case (Python)
```
data = ["ACG", "GCA"]
feats = string_features(data) # this automatically maps the given strings to `SGVector<char>` internally
svm.train(feats)
feats.get_strings() # returns `["ACG", "GCA"]` # would throw a runtime error for other template parameters of string features
feats.get_vectors() # returns `np.array(["ACG", "GCA"], dtype=char)
```
in those lines

Reply:
@karlnapf To make sure I understand correctly, you meant
``` python
feats.get_vectors() # returns `np.array([['A', 'C', 'G'], ['G', 'C', 'A']], dtype=char)
```
right? I'm ok with doing that, but I somewhat don't like the asymmetry, it's a bit confusing, and don't completely get the point.
What's confusing is: if a function takes `std::vector<SGVector<char>>` as an argument, the user should provide `String[]`. But if a function returns it (might be `get` from `SGObject`), he should expect to get `char[][]`. If we are not to treat `SGVector<char>` differently, we should do it all the way.

Reply:
I meant that, yes, sorry!

Reply:
Sorry, I haven't been clear. The asymmetry is what I don't like either, and I also don't like special treatments (`SGVector<char>` as `String`).

So the two fixes would be: everything is a vector, no more strings internally. But then we can have user-facing methods where users can pass their strings directly to methods that expect arrays of chars. Maybe let's discuss this in irc?